File: rust/src/error_handler.rs
use log::Level;
use std::collections::HashMap;

use actix_web::{
    http::{header::ContentType, StatusCode},
    HttpResponse, ResponseError,
};

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Validation errors: `{0:?}`")]
    ValidationErrors(#[from] validator::ValidationErrors),

    #[error("Persistance error: `{0:?}`")]
    PersistanceError(#[from] sqlx::Error),
}

impl ResponseError for Error {
    fn error_response(&self) -> HttpResponse {
        match self {
            Error::ValidationErrors(validation_errors) => {
                let mut errors = HashMap::new();

                validation_errors
                    .field_errors()
                    .iter()
                    .for_each(|(&field, &validation_errors)| {
                        errors.entry(field).or_insert_with(Vec::new).append(
                            &mut validation_errors
                                .iter()
                                .filter_map(|e| e.message.as_deref())
                                .collect(),
                        )
                    });

                log::warn!("{:#?}", self);
                HttpResponse::BadRequest().json(errors)
            }
            Error::PersistanceError(sqlx_error) => {
                let (status, message, level) = match sqlx_error {
                    // https://www.postgresql.org/docs/current/errcodes-appendix.html
                    sqlx::Error::Database(error) => match error.code().as_deref() {
                        // unique violation
                        Some("23505") => (
                            StatusCode::CONFLICT,
                            error.message().to_owned(),
                            Level::Warn,
                        ),

                        // integrity constraint violation | foreign key violation
                        Some("23000" | "23503") => (
                            StatusCode::BAD_REQUEST,
                            error.message().to_owned(),
                            Level::Warn,
                        ),

                        _ => (
                            StatusCode::INTERNAL_SERVER_ERROR,
                            error.message().to_owned(),
                            Level::Error,
                        ),
                    },
                    error => (
                        StatusCode::INTERNAL_SERVER_ERROR,
                        error.to_string(),
                        Level::Error,
                    ),
                };

                log::log!(level, "{}", message);
                HttpResponse::build(status)
                    .set(ContentType::plaintext())
                    .body(message)
            }
        }
    }
}

File: java/src/main/java/club/marzipan/ErrorHandler.java
package club.marzipan;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.hibernate.JDBCException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestController;

@RestController
@ControllerAdvice
public class ErrorHandler {

    private static final Logger logger = LoggerFactory.getLogger(ErrorHandler.class);

    private static final MediaType plaintext = MediaType.valueOf("text/plain; charset=utf-8");

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, List<String>>> handleValidationExceptions(
            MethodArgumentNotValidException methodArgumentNotValidException) {

        logger.warn(methodArgumentNotValidException.getMessage());
        Map<String, List<String>> errors = new HashMap<>();

        methodArgumentNotValidException.getBindingResult().getFieldErrors().forEach((field_error) -> {
            List<String> messages = errors.computeIfAbsent(field_error.getField(), field -> new ArrayList<>());
            messages.add(field_error.getDefaultMessage());
        });

        return ResponseEntity.badRequest().body(errors);
    }

    @ExceptionHandler(JDBCException.class)
    public ResponseEntity<String> handleSQLException(JDBCException jdbcException) {

        SQLException sqlException = jdbcException.getSQLException();
        HttpStatus status;

        switch (sqlException.getSQLState()) {
            // https://www.postgresql.org/docs/current/errcodes-appendix.html

            case "23505": // unique violation
                status = HttpStatus.CONFLICT;
                logger.warn(sqlException.getMessage());
                break;

            case "23000": // integrity constraint violation
            case "23503": // foreign key violation
                status = HttpStatus.BAD_REQUEST;
                logger.warn(sqlException.getMessage());
                break;

            default:
                status = HttpStatus.INTERNAL_SERVER_ERROR;
                logger.error(sqlException.getMessage());
        }

        return ResponseEntity.status(status).contentType(plaintext).body(sqlException.getMessage());
    }

}
